src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonOutputFormatter.cs

```diff
@@ -41,6 +41,12 @@ internal static SystemTextJsonOutputFormatter CreateFormatter(JsonOptions jsonOp
             };
         }
 
+        if (!jsonSerializerOptions.IsReadOnly &&
+            jsonSerializerOptions.TypeInfoResolver != null)
+        {
+            jsonSerializerOptions.MakeReadOnly();
+        }
```

Comment:
Is this really necessary? In this case we're not even storing the JsonTypeInfo. We resolve it later right before we serialize which should make the options read-only anyway, right?

It also feels a little odd that we go out of our way to not mutate the options by making a copy before configuring the default encoder, only to mutate it here.

I get that making the options read-only is a bit different than adding options, but it's still weird that we only lock down the original options instance rather than a copy if you happen to set a non-null encoder.

Reply:
Yeah, your are right. Making it readonly here make almost no difference. Let me remove it.

Reply:
On the contrary, I think it makes sense to do this early on and fail fast if, say, the `JsonSerializerOptions` instance is misconfigured for NativeAOT (i.e. `TypeInfoResolver` is null). I would suggest replacing the above with just
```C#
jsonSerializerOptions.MakeReadOnly(); // throws if `TypeInforesolver` is null.
```